### PR TITLE
[Flux] Further optimize FSDP performance 

### PR DESCRIPTION
**Pull request by wwwjn**
We used 3 FSDP root in current implementation: For CLIP encoder, T5 Encoder, FLUX model

Current Encoders (before optimization): 
Screenshot 2025-05-29 at 2 31 12 PM

Current FLUX model (before optimization): 
Screenshot 2025-05-29 at 12 13 12 PM

Opportunities from profiler: 
1. Optimize FLUX model : Increase local batch size / wrap multiple layers into one FSDP block.
2. Optimize CLIP encoder: Current CLIP encoder has huge communication overhead. We might now need to wrap CLIP encoder using FSDP
3. Optimize “command buffer full” which is blocking more computations: remove kernel launches.  If the kernel luanchs are from fsdp, we can fix it. if it's from T5/CLIP, we can consider torch.compile
     - There's a long all_gather() during the command buffer full, need to figure out
     - After setup a barrier after dataloader, there's still a long all_gather and "comm buffer full" for the first block of T5 encoder.
4. Optimize `to()` copy: For FLUX model, the inputs are generated random noise (called latents in our code), and then call  to(). We should try generating the random noise on GPU directly.

Thank @weifengpy  for help identifying issues!

### CODE REVIEW

**Comment by weifengpy** [comment]
> why is command buffer full a problem? it does not affect GPU execution -- it's actually good that the CPU is running ahead

If gpu kernels and cpu overhead are evenly distributed, I agree "command buffer full" does not matter

when not evenly distributed, I treat "command buffer full" as equivalent to cpu sync point. "command buffer full" resets the cpu thread headroom. attached profiler below, "Encoders (forward only)" is blcoked due to "command buffer full". "Flux model forward" could have launched kernels sooner. Because of the blocking, it become cpu bound now

Screenshot 2025-05-29 at 8 47 55 PM

another example is casting 1024 input parameters in `fully_shard(mp_policy)`:  https://github.com/pytorch/pytorch/issues/154272

File: torchtitan/experiments/flux/train.py [hunk 1]
```diff
@@ -112,32 +112,33 @@ def train_step(self, input_dict: dict[str, torch.Tensor], labels: torch.Tensor):
 
         bsz = labels.shape[0]
```
Comment (fegin) [inline torchtitan/experiments/flux/train.py:113]: Why is `labels` on CPU? Shouldn't it be on GPU? Then you won't need `with torch.device(self.device)` and the extra indention.

File: torchtitan/experiments/flux/train.py [hunk 2]
```diff
@@ -62,15 +62,15 @@ def __init__(self, job_config: JobConfig):
...
-
-        self.clip_encoder = FluxEmbedder(
-            version=job_config.encoder.clip_encoder,
-            random_init=job_config.training.test_mode,
-        ).to(device=self.device, dtype=self._dtype)
-        self.t5_encoder = FluxEmbedder(
-            version=job_config.encoder.t5_encoder,
-            random_init=job_config.training.test_mode,
-        ).to(device=self.device, dtype=self._dtype)
+        with torch.device(self.device):
```
Comment (fegin) [inline torchtitan/experiments/flux/train.py]: Is this necessary since there are already `to(device=self.device)`.

Comment (wwwjn) [reply torchtitan/experiments/flux/train.py]: My understanding is, if we use `with torch.device(self.device):`, the FluxEmbedder will be loaded to self.device (cuda) directly. If we initialize it and then call to(), it will first load it on CPU, then copy to self.device. Is this correct?

Comment (fegin) [reply torchtitan/experiments/flux/train.py]: Yes. But I'm not so sure if we want to initialize the model on CPU or GPU first. Our strategy is to do `with torch.device("meta"):`, then apply the parallelisms and finally move the sharded model directly to GPU.

Either way, calling both `with torch.device()` and `to()` is not necessary. You should just stick one.

**Comment by wwwjn** [comment]
For opportunity 1 ✅:
- Increase batch size, still wrap each block into one FSDP layer 

For opportunity 2 ✅:
- We don't warp CLIP model with FSDP anymore, because the CLIP encoder has little computation but large communication overhead. And the CLIP model size is small.

For opportunity 3:
- The long all_gather wait is because one dataloader out of the 8 ranks runs super slow. From the profiler(with_stack=True), it was slow partially because download dataset from internet.
- After adding a barrier between before train_step(), the long all_gather diminished. After optimization 1 (increase batch size), the training process is not CPU-bounded anymore, so "command buffer full" is not concerning. ✅
Screenshot 2025-06-02 at 3 38 17 PM

For opportunity 4 ✅: 
- It happens after randn_like(labels), so it could be `timesteps = torch.rand((bsz,)).to(self.device)`. After generating the `noise` and `labels` on GPU, the time was saved.

**Comment by wwwjn** [comment]
> @wwwjn Why isn't labels moved to GPU first?

The `labels` are already on GPU at the beginning of `train_step()`.  I think the "to copy" was made by the line followed by `rankn_like()`, which is `timesteps = torch.rand((bsz,)).to(self.device)`

**Comment by wwwjn** [comment]
Before the optimization: 
- total time for one step is ~900ms (local batch size = 4)
- Screenshot 2025-06-02 at 9 58 37 PM


After optimization: 
- total time for one step is ~7s (local batch size = 64)
- Screenshot 2025-06-02 at 9 59 32 PM

File: torchtitan/experiments/flux/train_configs/debug_model.toml [hunk 3]
```diff
@@ -33,8 +33,7 @@ warmup_steps = 1  # 10% warmup steps
 decay_ratio = 0.0  # no decay, stay stable during training
 
 [training]
-batch_size = 4
-seq_len = 512
+batch_size = 32
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/train_configs/debug_model.toml]: Maybe use a smaller batch size for debug model training?
Since the model is smaller, it's likely a smaller batch size (with T5 seq_len 256) still won't expose FSDP comms.

File: torchtitan/experiments/flux/parallelize_flux.py [hunk 4]
```diff
@@ -136,18 +137,15 @@ def parallelize_encoders(
         }
         if job_config.training.enable_cpu_offload:
             fsdp_config["offload_policy"] = CPUOffloadPolicy()
-        # FSDP for encoder blocks
-        for block in clip_model.hf_module.text_model.encoder.layers:
-            fully_shard(block, **fsdp_config)
-        fully_shard(clip_model, **fsdp_config)
 
+        # NOTE: only apply FSDP to the T5 encoder, not the CLIP text encoder
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/parallelize_flux.py]: please also mention the reason, i.e. CLIP is small and compute / comm ratio is low

**Review by tianyu-l — APPROVED** [review]
LGTM. Please address final comments before merge.
